server/src/main/java/org/opensearch/index/mapper/StarTreeMapper.java

```diff
@@ -0,0 +1,430 @@
+import org.opensearch.index.compositeindex.datacube.startree.StarTreeFieldConfiguration;
+import org.opensearch.index.compositeindex.datacube.startree.StarTreeIndexSettings;
+import org.opensearch.index.query.QueryShardContext;
+import org.opensearch.search.lookup.SearchLookup;
+
+import java.util.ArrayList;
+import java.util.LinkedHashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * A field mapper for star tree fields
+ *
+ * @opensearch.experimental
+ */
+@ExperimentalApi
+public class StarTreeMapper extends ParametrizedFieldMapper {
+    public static final String CONTENT_TYPE = "star_tree";
+    public static final String CONFIG = "config";
+    public static final String MAX_LEAF_DOCS = "max_leaf_docs";
+    public static final String SKIP_STAR_NODE_IN_DIMS = "skip_star_node_creation_for_dimensions";
+    public static final String BUILD_MODE = "build_mode";
+    public static final String ORDERED_DIMENSIONS = "ordered_dimensions";
+    public static final String METRICS = "metrics";
+    public static final String NAME = "name";
+    public static final String TYPE = "type";
+    public static final String STATS = "stats";
+
+    @Override
+    public ParametrizedFieldMapper.Builder getMergeBuilder() {
+        return new Builder(simpleName(), objBuilder).init(this);
+
+    }
+
+    /**
+     * Builder for the star tree field mapper
+     *
+     * @opensearch.internal
+     */
+    public static class Builder extends ParametrizedFieldMapper.Builder {
+        private ObjectMapper.Builder objbuilder;
+        private static final Set<Class<? extends Mapper.Builder>> ALLOWED_DIMENSION_MAPPER_BUILDERS = Set.of(
+            NumberFieldMapper.Builder.class,
+            DateFieldMapper.Builder.class
+        );
+        private static final Set<Class<? extends Mapper.Builder>> ALLOWED_METRIC_MAPPER_BUILDERS = Set.of(NumberFieldMapper.Builder.class);
+
+        @SuppressWarnings("unchecked")
+        private final Parameter<StarTreeField> config = new Parameter<>(CONFIG, false, () -> null, (name, context, nodeObj) -> {
+            if (nodeObj instanceof Map<?, ?>) {
+                Map<String, Object> paramMap = (Map<String, Object>) nodeObj;
+                int maxLeafDocs = XContentMapValues.nodeIntegerValue(
+                    paramMap.get(MAX_LEAF_DOCS),
+                    StarTreeIndexSettings.STAR_TREE_DEFAULT_MAX_LEAF_DOCS.get(context.getSettings())
+                );
+                if (maxLeafDocs < 1) {
+                    throw new IllegalArgumentException(
+                        String.format(Locale.ROOT, "%s [%s] must be greater than 0", MAX_LEAF_DOCS, maxLeafDocs)
+                    );
+                }
+                paramMap.remove(MAX_LEAF_DOCS);
+                Set<String> skipStarInDims = new LinkedHashSet<>(
```

Comment:
nit: do we need LinkedHashSet? Set.of() should also work?

Reply:
No `set.of` doesn't work, we specifically need a `LinkedHashSet` to maintain the order.  So the idea is that mapper first gets called to parse user input as below :
```
 documentMapper = documentParser.parse(type, entry.getValue());
```
Then its again called as part of 'merge' flow which merges all the properties as part of following code: 
```
internalMerge(documentMapper, reason);
```
During this flow,  we check if the user provided input order is maintained when we do merge - as part of following method 
```
assert results.values().stream().allMatch(this::assertSerialization);
```

So its essential to maintain order , for which we need to use LinkedHashSet